Make Hotcue RGB colors configurable

Holzhaus opened the pull request:
Follow up to #2520. This adds multiple predefined color palettes and a preferences page to select one.

Holzhaus committed:
util/color/colorpalette: Add additional predefined color palettes

Holzhaus committed:
preferences: Add colors dialog to preferences menu

Holzhaus committed:
util/color/predefinedcolorpalettes: Add Serato DJ Pro/Lite Hotcue palettes

daschuer commented on src/util/color/predefinedcolorpalettes.cpp:
```diff
@@ -0,0 +1,160 @@
+constexpr mixxx::RgbColor kTraktorProTrackColorViolet(0xA669FF);
+constexpr mixxx::RgbColor kTraktorProTrackColorMagenta(0xFE55EA);
+
+constexpr mixxx::RgbColor kSeratoDJProTrackColorGrey1(0x333333);
```
Are these the visual colors or the colors stored in the track metadata? 
Please add a comment to the source code.

Holzhaus replied:
Visual. I'll add a comment.

Holzhaus committed:
preferences: Add Color Palette Editor

Holzhaus committed:
preferences/colorpalettesettings: Prevent built-in palette name collisions

Holzhaus committed:
preferences/dialog/dlgprefcolors: Show custom palettes in comboboxes

Holzhaus committed:
preferences/dialog: Move auto_hotcue_colors from deck to colors page

Holzhaus committed:
preferences/colorpalettesettings: Fix loading predefined color palettes

Holzhaus committed:
util/color/colorpalette: Add support for special hotcue color indices

Holzhaus committed:
util/color/predefinedcolorpalettes: Add Serato DJ Pro hotcue indices

Holzhaus committed:
preferences/colorpalettesettings: Read/Write hotcue indices from/to config

Holzhaus committed:
preferences/colorpaletteeditor: Add support for hotcue color indices

Holzhaus committed:
preferences/colorpaletteeditor: Get rid of ColorPaletteEditorTableView

Holzhaus committed:
preferences/colorpaletteeditormodel: Remove unused function

Holzhaus committed:
preferences/colorpaletteeditor: Improve button update code

Holzhaus committed:
preferences/colorpaletteeditor: Use nicer buttons

Holzhaus commented:
The Palette Editor was less work than I thought. Here's a little preview:
![Hotcue Palette Editor Preview](https://user-images.githubusercontent.com/1834516/76096360-3f657b00-5fc6-11ea-93e7-0cd4396c811d.gif)

Holzhaus commented:
This is lacking a nice icon for the the preference page, otherwise this is ready to review.

Holzhaus commented:
@ronso0 Do you want to make an icon? Maybe a paintbrush or something like that?

daschuer commented:
How about something similar to this: https://emojipedia.org/google/android-4.3/artist-palette/

daschuer commented:
Thank you for the nice little preview.
Did you consider to add an edit mode to hide edit features?

The GUI suits for users who wants to tweak the palette. However this is a rare use case.

We should stream line the GUI for the common task to switch to Serato colors.

daschuer commented:
The hotcue default color selection can be improved. Did you consider to use the combo box from my proposal?

Do we need a way to go back to default orange once another color has been selected?

ronso0 commented:
> @ronso0 Do you want to make an icon? Maybe a paintbrush or something like that?

Hehee, I was already looking for one.. :)
Reagrding the icon size in the Pref list I think a simple paint brush will do, or even just a brush tip.
A palette at that size is not clear and somehow looks like the GNome icon.

![color_palette_icon](https://user-images.githubusercontent.com/5934199/76101156-3b3d5b80-5fce-11ea-92a2-d91b98a321f0.jpg)

![color_palette_icon_s](https://user-images.githubusercontent.com/5934199/76101213-57d99380-5fce-11ea-8d4c-01a91e536bf6.jpg)

Holzhaus commented:
> Did you consider to add an edit mode to hide edit features?
>
> The GUI suits for users who wants to tweak the palette. However this is a rare use case.

We could remove the palette editor from the preferences page and instead add an "edit palettes" button that opens the palette editor ins a separate window. Is that what you had in mind?

Holzhaus commented:
> The hotcue default color selection can be improved. Did you consider to use the combo box from my proposal?

If you want to use a single color, you can just edit the palette and but a "1" next to the color you want (and remove all other number from the "Assign to Hotcue" column). That will color all new cues in that color. 

> Do we need a way to go back to default orange once another color has been selected?

You could just re-add orange to the palette and then do as I described above.

I think this allows great flexibility when configuring palettes. IMHO we should even drop the dedicated auto hotcue colors option and make that the default. If you want to make all new cues a single color, you can just configure the palette to your liking and only auto-assign a single color.

Holzhaus committed:
util/color/predefinedcolorpalettes: Add comments

Holzhaus committed:
util/color/predefinedcolorpalettes: Add Serato DJ Intro Hotcue Palette

ronso0 commented:
I made an svg from the brush tip.
Does this look okay to everyone?
![ic_preferences_colors svg_24](https://user-images.githubusercontent.com/5934199/76204603-fe08e180-61f8-11ea-849e-c2a83efec157.png)
![ic_preferences_colors svg_32](https://user-images.githubusercontent.com/5934199/76204604-fe08e180-61f8-11ea-89fd-dbd8f077ad61.png)
[ic_preferences_colors.svg.png.zip](https://github.com/mixxxdj/mixxx/files/4306050/ic_preferences_colors.svg.png.zip)

ywwg commented:
>  Does this look okay to everyone?

that seems fine -- the source image you used for this was an open source SVG, right?

ronso0 commented:
I created the svg by looking a some raster graphics.
Now it's opensource :)

ywwg commented:
> I created the svg by looking a some raster graphics.

IANAL but sgtm

ronso0 commented:
> I created the svg by looking a some raster graphics.

...and tweaking the design for the size we need in the Pref window.
This build errors, so I'll test the icon in master.
I'll be back with the final version then.

uklotzde commented on src/controllers/colormapperjsproxy.h:
```diff
@@ -0,0 +1,27 @@
+        delete m_colorMapper;
+    };
+
+  public slots:
```
Instead of slots we should use the Q_INVOKABLE macro:

https://doc.qt.io/qt-5/qobject.html#Q_INVOKABLE

Slots are invoked from the event loop and therefore don't return values. The is just a semantic difference, the actual, generated implementation will probably be identical.

Holzhaus commented:
I'll mark this as WIP until #2520 has been merged.

Holzhaus commented:
> I made an svg from the brush tip.
> Does this look okay to everyone?
> ![ic_preferences_colors svg_24](https://user-images.githubusercontent.com/5934199/76204603-fe08e180-61f8-11ea-849e-c2a83efec157.png)
> ![ic_preferences_colors svg_32](https://user-images.githubusercontent.com/5934199/76204604-fe08e180-61f8-11ea-89fd-dbd8f077ad61.png)
> [ic_preferences_colors.svg.png.zip](https://github.com/mixxxdj/mixxx/files/4306050/ic_preferences_colors.svg.png.zip)

Thanks, looks good to me. Unfortunately you uploaded a PNG image, but all the other icons in `res/images/preferences` are SVGs.

ronso0 commented:
oopsy.. anyway, that icon turned out to not work well at 16x16 px in the actual Pref list, it's simply too small and thus quite blurry.
I tried another brush shape which works better at 16px, and I hope can finalize that soonish.
have a look:
![ic_preferences_colors svg_16e](https://user-images.githubusercontent.com/5934199/76308408-ffa2da00-62ca-11ea-927b-e140377750d1.png)

ronso0 committed:
add brush icon for Preferences > Colors

ronso0 commented:
Looks good now

@100% (abused **Broadcast** because I couldn't build this branch)
![image](https://user-images.githubusercontent.com/5934199/76641237-8f59ba00-6551-11ea-9aac-3ddb885c5ab0.png)

the icons is [here](825909091b)

Holzhaus commented:
@ronso0 can you post the exact error message? AFAIK this should build fine.

Holzhaus commented:
Thanks for taking care of the Icon Btw, but I'm getting a 404 when I open your link.

ronso0 commented:
> Thanks for taking care of the Icon Btw, but I'm getting a 404 when I open your link.

strange, I tested it.
this should work https://github.com/ronso0/mixxx/commit/825909091bd27c5d84cfe73197c373c62ce143be

ronso0 commented:
> @ronso0 can you post the exact error message? AFAIK this should build fine.

It was a few days back and I didn't try again just for the icon.
I'll do so tomorrow night and let you know.

Holzhaus committed:
res/mixxx.qrc: Add ic_preferences_colors.svg to QRC file

ywwg commented on src/preferences/colorpaletteeditor.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    connect(m_pTableView,
+            &QTableView::customContextMenuRequested,
+            [this](const QPoint& pos) {
```
please only only use lambdas for very short, trivial functions.  This is long enough that it should be a separate handler.

ywwg replied:
(There are a lot of handlers here, nearly all of them are fairly long and complicated. If they are broken into functions then we might actually be able to test them)

ywwg commented on src/preferences/colorpaletteeditor.cpp:
```diff
@@ -0,0 +1,199 @@
+            [this](const QString& text) {
+                bool bPaletteIsReadOnly = false;
+                bool bPaletteExists = false;
+                foreach (const ColorPalette& palette, mixxx::PredefinedColorPalettes::kPalettes) {
```
I believe you can use `for ( x : y )` instead of foreach now.

ywwg commented on src/preferences/colorpaletteeditormodel.h:
```diff
@@ -0,0 +1,45 @@
+
+#include "util/color/colorpalette.h"
+
+class ColorPaletteEditorModel : public QStandardItemModel {
```
please include documentation comments with all new classes

ywwg reviewed (changes requested):
first round of comments.  Thanks again for doing all this

Holzhaus committed:
preferences/colorpaletteeditor: Move lambda functions to private slots

Holzhaus committed:
preferences/colorpaletteeditor: Replace foreach() with for (x : y)

Holzhaus committed:
preferences/colorpaletteeditor(model): Add class comments to headers

Holzhaus commented:
Anything else or can we merge this?

uklotzde commented on src/preferences/colorpaletteeditormodel.h:
```diff
@@ -0,0 +1,48 @@
+        emit dirtyChanged(m_bDirty);
+    }
+
+    bool isDirty() {
```
Both functions should be const

uklotzde commented on src/preferences/colorpalettesettings.cpp:
```diff
@@ -47,6 +75,12 @@ void ColorPaletteSettings::setColorPalette(const QString& name, const ColorPalet
         return;
     }
 
+    foreach (const ColorPalette& palette, mixxx::PredefinedColorPalettes::kPalettes) {
```
We should prefer C++ range-based for loops. Applies to multiple occurrences.

uklotzde commented on src/preferences/dialog/dlgprefcolors.h:
```diff
@@ -0,0 +1,32 @@
+    void loadSettings();
+    void loadPaletteIntoEditor(const ColorPalette& palette);
+
+    // Pointer to config object
```
This comment adds no value. All those members that are initialized once in the constructor could also be declared as const.

uklotzde commented on src/preferences/colorpaletteeditor.h:
```diff
@@ -0,0 +1,48 @@
+    Q_OBJECT
+  public:
+    ColorPaletteEditor(QWidget* parent = nullptr);
+    void setConfig(UserSettingsPointer pConfig) {
```
- Why not configure() or initialize()? This functions is supposed to do more than just setting a member.
- Maybe even add DEBUG_ASSERT(!m_pConfig) if it is only supposed to be called once.
- Better hide the actual implementation in the .cpp file

Holzhaus replied:
OK. That function is just a crutch because I don't know how to pass custom args to the constructor of Qt Designer widgets.

uklotzde reviewed (commented):
Just some minor nitpicks

Holzhaus committed:
preferences/colorpaletteeditormodel: Make isDirty/isEmpty const

Holzhaus committed:
preferences/colorpalettesettings: Use range-based for loops

Holzhaus committed:
preferences/dialog/dlgprefcolors: Fixed name and constness of config ptr

ywwg reviewed (approved):
lgtm

Holzhaus committed:
preferences/colorpaletteeditor: Improve widget initialization

Holzhaus commented:
@uklotzde Done.

uklotzde commented:
Debug assertion on shutdown (some line numbers may vary, merged into my personal development branch):

```
Debug [Main]: 910 ms deleting DlgPreferences
DEBUG ASSERT: "!m_ptr || m_ptr->parent()" in function parented_ptr<ColorPaletteEditorModel>::~parented_ptr() [T = ColorPaletteEditorModel] at /tmp/mixxx/src/util/parented_ptr.h:31

Thread 1 "mixxx" received signal SIGABRT, Aborted.
```

```
#7  0x00000000004d0e4e in mixxx_debug_assert(char const*, char const*, int, char const*)
    (assertion=0x107e320 "!m_ptr || m_ptr->parent()", file=0x107e33a "/tmp/mixxx/src/util/parented_ptr.h", line=31, function=0x10c8cc4 "parented_ptr<ColorPaletteEditorModel>::~parented_ptr() [T = ColorPaletteEditorModel]") at /tmp/mixxx/src/util/assert.h:11
#8  0x0000000000797e20 in parented_ptr<ColorPaletteEditorModel>::~parented_ptr() (this=0x5673e18) at /tmp/mixxx/src/util/parented_ptr.h:31
#9  0x000000000077d5e3 in ColorPaletteEditor::~ColorPaletteEditor() (this=0x5673dc0) at /tmp/mixxx/src/preferences/colorpaletteeditor.h:14
#10 0x000000000077d64c in ColorPaletteEditor::~ColorPaletteEditor() (this=0x5673dc0) at /tmp/mixxx/src/preferences/colorpaletteeditor.h:14
#11 0x00007ffff42f584c in QObjectPrivate::deleteChildren() () at /lib64/libQt5Core.so.5
#12 0x00007ffff6e306b9 in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#13 0x00007ffff6f2ba9d in QGroupBox::~QGroupBox() () at /lib64/libQt5Widgets.so.5
#14 0x00007ffff42f584c in QObjectPrivate::deleteChildren() () at /lib64/libQt5Core.so.5
#15 0x00007ffff6e306b9 in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#16 0x0000000000559018 in DlgPreferencePage::~DlgPreferencePage() (this=0x5675030) at /tmp/mixxx/src/preferences/dlgpreferencepage.cpp:8
#17 0x0000000000b592d1 in DlgPrefColors::~DlgPrefColors() (this=0x5675030) at /tmp/mixxx/src/preferences/dialog/dlgprefcolors.cpp:33
#18 0x0000000000b5931c in DlgPrefColors::~DlgPrefColors() (this=0x5675030) at /tmp/mixxx/src/preferences/dialog/dlgprefcolors.cpp:32
#19 0x00007ffff42f584c in QObjectPrivate::deleteChildren() () at /lib64/libQt5Core.so.5
#20 0x00007ffff6e306b9 in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#21 0x00007ffff6e308ad in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#22 0x00007ffff42f584c in QObjectPrivate::deleteChildren() () at /lib64/libQt5Core.so.5
#23 0x00007ffff6e306b9 in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#24 0x00007ffff6f9289d in QScrollArea::~QScrollArea() () at /lib64/libQt5Widgets.so.5
#25 0x00007ffff42f584c in QObjectPrivate::deleteChildren() () at /lib64/libQt5Core.so.5
#26 0x00007ffff6e306b9 in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#27 0x00007ffff6fa898d in QStackedWidget::~QStackedWidget() () at /lib64/libQt5Widgets.so.5
#28 0x00007ffff42f584c in QObjectPrivate::deleteChildren() () at /lib64/libQt5Core.so.5
#29 0x00007ffff6e306b9 in QWidget::~QWidget() () at /lib64/libQt5Widgets.so.5
#30 0x00000000005094cd in DlgPreferences::~DlgPreferences() (this=0x432e130) at /tmp/mixxx/src/preferences/dialog/dlgpreferences.cpp:183
#31 0x000000000050954c in DlgPreferences::~DlgPreferences() (this=0x432e130) at /tmp/mixxx/src/preferences/dialog/dlgpreferences.cpp:171
#32 0x00000000004db381 in MixxxMainWindow::finalize() (this=0x7fffffffcfd0) at /tmp/mixxx/src/mixxx.cpp:745
```

uklotzde commented:
I also see a warning during startup:

```
Debug [Main]: DbConnectionPooled - Found thread-local database connection "MIXXX-1" QSqlDatabase(driver="QSQLITE", database="file:///home/uk/.mixxx/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=true)
Warning [Main]: Color Palette "Mixxx Hotcue Colors" is a built-in palette, not writing to config!
Warning [Main]: Color Palette "Mixxx Hotcue Colors" is a built-in palette, not writing to config!
Debug [Main]: Key plugin ID: "qm-keydetector:2"
```

This should not happen. Either the log level is wrong or we should handle this case gracefully.

Holzhaus committed:
preferences/colorpalettesettings: Print debug message when skipping write

Holzhaus committed:
preferences/colorpaletteeditor: Don't use parented_ptr for model

Since the parented_ptr approach caused crashes on shutdown, I switched
to a regular pointer and set the QTableView as the parent object of
model. According to the Qt docs, this should transfer ownership to the
table view when setModel() is called.

uklotzde reviewed (commented):
Please revert the last commit and keep using parented_ptr. Only the parent `m_pTableView` was missing:

```
m_pModel(make_parented<ColorPaletteEditorModel>(m_pTableView))
```

Holzhaus committed:
Revert "preferences/colorpaletteeditor: Don't use parented_ptr for model"

This reverts commit de013aad58ce129fc00af934fa6ea7b0f11cd1b5.

Holzhaus committed:
preferences/colorpaletteeditor: Pass m_pTableView to model

uklotzde commented:
Good job, Jan!

LGTM

JaviVilarroig commented:
@Holzhaus It seems I was too slow. Still slow with git. I have just managed to compile this PR. Still interested in any testing from my side?

ywwg commented:
it's ok to test still, and if you catch any problems we'll address them for sure!

Holzhaus commented:
Yup, but it make sense to test this on `master`. That way you catch problems that were introduced by the merge itself as well.

JaviVilarroig commented:
OK. Now that I know how to do it I will move to master :)
Anything special to test? O do I just "play" a little bit with the colors?

Holzhaus commented:
No just play around and check if you can define custom cue palettes. I just found a little bug in master tho, so better test #2586.
